pkg/generic/meta.go

```diff
@@ -0,0 +1,60 @@
+package generic
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/label"
+	"github.com/giantswarm/microerror"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type OwnerClusterGetter func(metav1.ObjectMetaAccessor) (capi.Cluster, bool, error)
+
+func TryGetClusterName(object metav1.ObjectMetaAccessor) (string, bool) {
+	if object.GetObjectMeta() == nil || object.GetObjectMeta().GetLabels() == nil {
+		return "", false
+	}
+	labels := object.GetObjectMeta().GetLabels()
+
+	// First let's try to get CAPI cluster name label.
+	clusterName := labels[capi.ClusterLabelName]
+	if clusterName == "" {
+		// CAPI cluster name label not found, now let's try GS cluster ID label, which is basically
+		// the same thing.
+		clusterID := labels[label.Cluster]
+		if clusterID != "" {
+			// We found GS cluster ID, this is our cluster name.
+			clusterName = clusterID
+		}
+	}
+
+	return clusterName, clusterName != ""
+}
+
+func GetOwnerCluster(ctx context.Context, ctrlClient client.Client, object metav1.ObjectMetaAccessor) (capi.Cluster, bool, error) {
```

Comment:
I'd maybe use `client.Reader` here instead of `client.Client` so it would be possible to use caching client if desired.